neqo-http3/src/send_message.rs

```diff
@@ -0,0 +1,270 @@
+        }
+    }
+
+    pub fn is_sending_closed(&self) -> bool {
+        match self.state {
+            SendMessageState::Initialized { fin, .. }
+            | SendMessageState::SendingInitialMessage { fin, .. } => fin,
+            SendMessageState::SendingData => false,
+            _ => true,
+        }
+    }
+
+    pub fn done(&self) -> bool {
+        matches!(self.state, SendMessageState::Closed)
+    }
+
+    pub fn is_state_sending_data(&self) -> bool {
+        matches!(self.state, SendMessageState::SendingData)
+    }
+
+    fn ensure_response_encoded(
+        &mut self,
+        conn: &mut Connection,
+        encoder: &mut QPackEncoder,
+    ) -> Res<()> {
+        if let SendMessageState::Initialized { headers, data, fin } = &self.state {
+            qdebug!([self], "Encoding headers");
+            let header_block = encoder.encode_header_block(conn, &headers, self.stream_id)?;
+            let hframe = HFrame::Headers {
+                header_block: header_block.to_vec(),
+            };
+            let mut d = Encoder::default();
+            hframe.encode(&mut d);
+            if let Some(buf) = data {
+                qdebug!([self], "Encoding data");
+                let d_frame = HFrame::Data {
+                    len: buf.len() as u64,
+                };
+                d_frame.encode(&mut d);
+                d.encode(&buf);
+            }
+
+            self.state = SendMessageState::SendingInitialMessage {
+                buf: d.into(),
+                fin: *fin,
+            };
+        }
+        Ok(())
+    }
+
+    pub fn send(&mut self, conn: &mut Connection, encoder: &mut QPackEncoder) -> Res<()> {
+        self.ensure_response_encoded(conn, encoder)?;
+
+        let label = if ::log::log_enabled!(::log::Level::Debug) {
+            format!("{}", self)
+        } else {
+            String::new()
+        };
+
+        if let SendMessageState::SendingInitialMessage { ref mut buf, fin } = self.state {
+            let sent = conn.stream_send(self.stream_id, &buf)?;
+            qtrace!([label], "{} bytes sent", sent);
+
+            if sent == buf.len() {
+                if fin {
+                    conn.stream_close_send(self.stream_id)?;
+                    self.state = SendMessageState::Closed;
+                    qtrace!([label], "done sending request");
+                } else {
+                    self.state = SendMessageState::SendingData;
+                    self.conn_events.data_writable(self.stream_id);
+                    qtrace!([label], "change to state SendingData");
+                }
+            } else {
+                let b = buf.split_off(sent);
+                *buf = b;
+            }
+        }
+        Ok(())
+    }
+
+    // TransactionClient owns headers and sends them. This method returns if
+    // they're still being sent. Request body (if any) is sent by http client
+    // afterwards using `send_request_body` after receiving DataWritable event.
```

Comment:
I didn't see it anywhere, but we should intercept the DataWritable event while this method returns true.

Reply:
We do already in connection_client.
The server side do not emit DataWritable events at the moment.